### ISSUE TITLE
Dangerous mixing of logical espressions with optional arguments

### ISSUE THREAD

**Issue by giacrossi** [pdebuyl/fortran_tester#17]
The Fortran 2003 standard clearly specify that there's not an established order for the operand evaluation of a logical expression: 

> It is not necessary for a processor to evaluate all of the operands of an expression, or to evaluate entirely each operand, if the value of the expression can be determined otherwise.

So the expression
`if (.not. present(fail) .or. (present(fail) .and. fail .eqv. .false.)) then`
could lead to unpredictable result.

What's the purpose of this expression? Could you clarify it in order to split it and at first evaluate if the optional argument is present?

**Comment by pdebuyl** [comment]
The principle is that the `fail` flag allows you to enable tests to fail. It does allow tests to pass though.

Whatever the usage of the flag, your comment holds: everything in the if clause will be evaluated and this is undesirable. If you feel like doing so, you can submit a PR to fix the issue.

If any user of the library wishes to comment on whether they use the `fail` flag, their feedback is welcome.

**Comment by szaghi** [comment]
Hi @pdebuyl 

I think there is a problem in the `if` operands, in particular

```fortran
(present(fail).and.fail.eqv..false.))
```
is problematic because the processor is allowed to evaluate `fail.eqv..false.` before `present(fail)` (or even to evaluate only `fail.eqv..false.`) with unpredictable results (thus some compilers can raise a warning/error if enough smart).

The fix is simple, split the operands like

```fortran
if (present(fail)) then
...
```

If no one will create a PR, I can try to do my best the next Monday.

Cheers

**Comment by pdebuyl** [comment]
Hi @szaghi thank you for checking in. I probably won't have time to do this by monday, I'll look at your PR if there is one. Else, I'll try to get rid of this problem. BTW, is anyone using the "fail" flag?

**Comment by szaghi** [comment]
@pdebuyl Hi Pierre, I am not sure, I have to check. It is possible that I am using it. Monday I'll check all :+1:

**Comment by pdebuyl** [comment]
For info, I have only one occurence of `fail = .true.` in my codes and it would be easy to remove it.

I actually use cmake with

    set_tests_properties(test_name PROPERTIES WILL_FAIL TRUE)

for such tests now.

**Comment by szaghi** [comment]
@pdebuyl 

Hi Pierre, I have just checked, I am not using the `fail` flag, for me it is safe to remove, but I am only one user.

Cheers

**Comment by giacrossi** [comment]
Hi, in the Quantum ESPRESSO FFTXlib tests the flag Is never used: I can create a PR, just let me know if you prefer to delete the flag or change the if structure. Cheers

**Comment by pdebuyl** [comment]
fixed by #18
